crates/dbsp/src/trace/layers/ordered/mod.rs

```diff
@@ -453,6 +453,108 @@ where
     }
 }
 
+impl<K, L, O> OrderedBuilder<K, L, O>
+where
+    K: Ord + Clone,
+    L: MergeBuilder,
+    O: OrdOffset,
+{
+    /// Like `copy_range`, but uses `fuel` to bound the amount of work.
+    ///
+    /// Builds at most `fuel` values plus values for one extra key.
+    /// If `fuel > 0` when the method returns, this means that the merge is
+    /// complete.
+    pub fn push_merge_fueled(
+        &mut self,
+        input1: (&<Self as Builder>::Trie, &mut usize, usize),
+        input2: (&<Self as Builder>::Trie, &mut usize, usize),
+        fuel: &mut isize,
+    ) {
+        let (source1, lower1, upper1) = input1;
+        let (source2, lower2, upper2) = input2;
+
+        let starting_updates = self.vals.keys();
+        let mut effort = 0isize;
+
+        // while both mergees are still active
+        while *lower1 < upper1 && *lower2 < upper2 && effort < *fuel {
+            self.merge_step((source1, lower1, upper1), (source2, lower2, upper2));
+            effort = (self.vals.keys() - starting_updates) as isize;
+        }
+
+        // Merging is complete; only copying remains.
+        if *lower1 == upper1 || *lower2 == upper2 {
+            // Limit merging by remaining fuel.
+            let mut remaining_fuel = *fuel - effort;
+            if remaining_fuel > 0 {
+                if *lower1 < upper1 {
+                    if remaining_fuel < 1_000 {
+                        remaining_fuel = 1_000;
+                    }
+                    *lower1 =
+                        self.copy_range_fueled(source1, *lower1, upper1, remaining_fuel as usize);
+                }
+                if *lower2 < upper2 {
+                    if remaining_fuel < 1_000 {
+                        remaining_fuel = 1_000;
+                    }
+                    *lower2 =
+                        self.copy_range_fueled(source2, *lower2, upper2, remaining_fuel as usize);
+                }
+            }
+        }
+
+        effort = (self.vals.keys() - starting_updates) as isize;
+
+        *fuel -= effort;
+
+        // if *fuel < -1_000_000 {
+        //     eprintln!("Massive deficit OrdKey::work: {}", fuel);
+        // }
+    }
+
+    /// Like `copy_range`, but uses `fuel` to bound the amount of work.
+    ///
+    /// Invariants:
+    /// - Copies at most `fuel` values plus values for one extra key.
+    /// - If `fuel` is greater than or equal to the number of values, in the
+    ///   `lower..upper` key range, copies the entire range.
+    pub fn copy_range_fueled(
```

Comment:
Is `.copy_range()` intensive enough for fuel limiting to matter? Wouldn't we be better off unconditionally coping the entire range over even if there's not technically enough fuel to do every element? Otherwise I can easily see situations where we've got a handful of elements left just kinda dangling that have to be re-addressed later on

Reply:
This is precisely the behavior that this PR fixes :) Turns out `copy_range` can be expensive when running unchecked.